apps/labs/posts/narwhals-pycapsule.md

```diff
@@ -0,0 +1,199 @@
+  >>> import duckdb
+  >>> import pandas as pd
+  >>> import pyarrow as pa
+  >>> hasattr(pl.DataFrame, '__arrow_c_stream__')
+  True
+  >>> hasattr(duckdb.DuckDBPyRelation, '__arrow_c_stream__')
+  True
+  >>> hasattr(pd.DataFrame, '__arrow_c_stream__')
+  True
+  >>> hasattr(pa.Table, '__arrow_c_stream__')
+  True
+  ```
+- The solution is totally agnostic to the exact input, as it respects a standardised interface.
+  We can verify this by trying to pass different dataframe libraries to the function:
+
+  ```python
+  >>> from pycapsule_demo import agnostic_sum_i64_column
+  >>> df_pl = pl.DataFrame({'a': [1,1,2], 'b': [4,5,6]})
+  >>> df_pd = pd.DataFrame({'a': [1,1,2], 'b': [4,5,6]})
+  >>> df_pa = pa.table({'a': [1,1,2], 'b': [4,5,6]})
+  >>> rel = duckdb.sql("select * from df_pd")
+  >>> agnostic_sum_i64_column(df_pl, column_name="a")
+  6
+  >>> agnostic_sum_i64_column(df_pd, column_name="a")
+  6
+  >>> agnostic_sum_i64_column(df_pa, column_name="a")
+  6
+  >>> agnostic_sum_i64_column(rel, column_name="a")
+  6
+  ```
+
+Like magic, our function works agnostically for any of these dataframes,
+without us having to write any specialised code to handle the subtle differences
+between them!
+
+> **_NOTE:_** How many times can `agnostic_sum_i64_column` be called? Strictly
+> speaking, the PyCapsule Interface only guarantees that it can be called once for
+> any given input. In practice, however, all implementations that we're aware of
+> (except for DuckDB) allow for it to be called repeatedly. Discussion about the
+> DuckDB implementation is ongoing at https://github.com/duckdb/duckdb/discussions/15536.
+
+If you found the above example a little daunting, you may be wondering if a simpler
+solution exists which you can develop entirely in Python-land. Enter: Narwhals.
+
+## Python solution: Narwhals
+
+Narwhals is a lightweight, dependency-free, and extensible compatibility layer between
+dataframe libraries functions. To use it to write a dataframe-agnostic function,
+you need to follow three steps:
+
+- Call `narwhals.from_native` on the user's object.
+- Use the [Narwhals API](https://narwhals-dev.github.io/narwhals/api-reference/).
+- If you want to return to the user an object of the same kind that they started with,
+  call `to_native`.
+
+In this case, a Narwhals version of `agnostic_sum_i64_column` looks like this:
+
+```python
+import narwhals as nw
+from narwhals.typing import IntoFrame
+
+
+def agnostic_sum_i64_column_narwhals(df_native: IntoFrame, column_name: str) -> int:
+    lf = nw.from_native(df_native).lazy()
+    schema = lf.collect_schema()
+    if column_name not in schema:
+        msg = f"Column '{column_name}' not found, available columns are: {schema.names()}."
+        raise ValueError(msg)
+    if (dtype := schema[column_name]) != nw.Int64:
+        msg = f"Column '{column_name}' is of type {dtype}, expected Int64"
+        raise TypeError(msg)
+    df = lf.collect()
+    return df[column_name].sum()
```

Comment:
What backend do you use to actually compute the sum?

Reply:
The native one (e.g. Polars in => polars compute, PyArrow in => PyArrow compute, pandas in => pandas compute, cuDF in cuDF compute)